### PR TITLE
Allow objective tensors to be more than 2D

### PR DESCRIPTION
**Pull request by jonilaserson**
This forces the sample weights to be a 2D tensor of shape (MINIBATCH_SIZE, 1).
As a result, the tensors y_pred and y_true retain their shape when the 'weighted' objective function is generated (line 66) and the user can generate his/her own objective functions on tensors larger than 2D.

### CODE REVIEW

**Comment by fchollet** [comment]
Objective tensors can at this time be any shape (2D, 3D, 4D). In fact we have tests for this (e.g. `tests/auto/test_tasks.py`).

Your PR, however, makes training fail for non-2D shapes. It does precisely the opposite of what the title says.

More specifically, it causes sequence-to-sequence training failure in `tests/auto/test_tasks.py`. Check it out.

**Comment by jonilaserson** [comment]
You are right that this fails the RNN tests. I haven't tested on it, and I
don't understand right now what my solution is doing to that model.

So my solution is not ready yet, but let me convince you that there is a
problem first.

Say my output is of shape (BATCH_SIZE, 2, 20, 20), and my labels (Y_train)
are of shape (BATCH_SIZE, 2, 20, 20) as well.
Wouldn't you say that in the objective function both y_pred and y_true
should be 4D?
The answer is that they are not.  They are 2D.
Do you agree to that observation?

For example, consider *mean_squared_error* objective function:

def mean_squared_error(y_true, y_pred):
    return T.sqr(y_pred - y_true).mean(axis=-1)

Say y_true and y_pred are 4D, as you say, don't you find it weird that the
mean is taken over just the last axis?  Wouldn't that make the returned
loss 3D?
It doesn't, and the reason is that there is another method ('weighted')
that flattens the output and the labels without asking you.

The tests in tests/auto/test_tasks.py are blind to this, because in the
test_img_clf test the output is flattened at the end.

If you want your loss to be MSE on the first output map and MAE on the
second output map, you will have to work hard to do in a custom objective
function.
With my proposed, you can write an objective as so:

def mean_error(y_true, y_pred):
    return T.sqr(y_pred[:,0] - y_true[:,0]).mean(axis=(1,2)) + \
           T.abs_(y_pred[:,1] - y_true[:,1]).mean(axis=(1,2))

So can we at least agree that there is a problem?

**Comment by fchollet** [comment]
> Wouldn't you say that in the objective function both y_pred and y_true
should be 4D?
> The answer is that they are not.  They are 2D.
> Do you agree to that observation?

Sure... they are reshaped to 2D at this point, but the learning behavior is still what is expected for e.g. sequence-to-sequence, or for image-to-image learning. Which is why these tasks do work fine in Keras.

> If you want your loss to be MSE on the first output map and MAE on the
second output map, you will have to work hard to do in a custom objective
function.

I am not sure what you mean here?

**Comment by jonilaserson** [comment]
On Aug 25, 2015 23:02, "François Chollet" <notifications@github.com> wrote:
>>
>> Wouldn't you say that in the objective function both y_pred and y_true
>> should be 4D?
>> The answer is that they are not. They are 2D.
>> Do you agree to that observation?
>
> Sure... they are reshaped to 2D at this point, but the learning behavior
is still what is expected for e.g. sequence-to-sequence, or for
image-to-image learning. Which is why these tasks do work fine in Keras.

But we are losing a ton of flexibility by flattening the 4D volumes.
Without the 4D volume, my custom objectives can't operate directly on the
rows/columns of the output (dimensions 3 and 4 in the 4D tensor) or treat
each output map (dimension 2) differently, as sometimes is the case.

>>
>> If you want your loss to be MSE on the first output map and MAE on the
>> second output map, you will have to work hard to do in a custom objective
>> function.
>
> I am not sure what you mean here?

If your output shape is (MINIBATCH_SIZE, 2, 27, 27), it is like saying that
for each sample you output two heatmaps of 27x27. Maybe the first map show
the confidence that a person was detected at each pixel, so you want binary
cross entropy on that map, and the second map has the person's predicted
age, so you want mae loss on that. But if you want to apply a different
loss on each heatmap you are stuck, because once the 4D tensor is flattened
(e.g. one vector per sample) you can't reconstruct those maps in the
objective function.

Do you not think this is an issue if we want keras to be easily extendable
for custom objectives?

**Comment by fchollet** [comment]
> Do you not think this is an issue if we want keras to be easily extendable
for custom objectives?

Sure, it makes it difficult to write custom loss functions that make use of columns in images, etc (you first have to un-flatten the inputs). However it is still doable (again, by un-flattening the inputs first). 

The thing is that *not* flattening the inputs to the loss functions would significantly complicate the system we have in place for loss / weighting / masking. Unless you have a simple solution.

If you have a solution that is better than the current setup, I will of course consider it. Until then, the choice is between making a very niche use case easier, or complicating a system that's already working for almost everyone.

**Comment by jonilaserson** [comment]
>
> Do you not think this is an issue if we want keras to be easily extendable
> for custom objectives?
>
> Sure, it makes it difficult to write custom loss functions that make use
> of columns in images, etc (you first have to un-flatten the inputs).
> However it is still doable (again, by un-flattening the inputs first).
>
> True, but this only works if the objective function knows exactly the
shape of your output. This will be the only objective in the bank that
depends on the input shape, and that's kind of ugly to have to manually
change the reshape values every time your input dimension changes.  Also,
one of the cool things about fully convolutional nets is that you don't
need to know the shape of the input/output in advance, and the same
compiled network can work on any image size. You will lose that ability to
train inputs of multiple sizes with such objective.


> The thing is that *not* flattening the inputs to the loss functions would
> significantly complicate the system we have in place for loss / weighting /
> masking. Unless you have a simple solution.
>
Are you sure it will "significantly complicate" the system?  My change was
only 3 lines of code.  I'm not familiar with the sides of keras that
involve weighting, termporal analysis, and masking, but what makes you
think this cannot be solved easily?

If you have a solution that is better than the current setup, I will of
> course consider it. Until then, the choice is between making a very niche
> use case easier, or complicating a system that's already working for almost
> everyone.
>
First, I completely disagree that it is a "very niche use case".  Almost
every DL paper I'm reading has a non-standard objective function, and it
would be a shame if such models would be a hassle to implement in keras.
For example, OverFeat, a very famous object detection vision framework,
uses such an objective.

Second, I think that the current working setup is not a feature, it's a bug.

For example, can you explain why this line
<https://github.com/fchollet/keras/blob/master/keras/models.py#L364> makes
the weights be the same size as the output labels?
self.weights = T.ones_like(self.y_train)
Shouldn't each sample has just *one* weight?

Also, does it make sense that self.loss receives y_pred and y_true as
unflattened tensors but the weighted train_loss gets them flattened?

To get the code back to a clean state and provide clarity, I think we need
to decide (and specify in the documentation) the following things:
1. What are y_pred and y_true, and what are their expected/required shapes
given the output of the network?
2. What are the sample weights?  What is the 'weights' variable shape?
3. How does masking interact with the weights?

**Comment by jonilaserson** [comment]
ping

**Comment by zhangtemplar** [comment]
@jonilaserson I want this feature too. I think current setting doesn't work well for regression problem.
